modules/standard/Allocators.chpl

```diff
@@ -0,0 +1,377 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/* Provides custom allocators for Chapel objects on the heap.
+
+  This module provides an interface to customize how Chapel objects are
+  allocated on the heap. This is opted into by changing calls to
+  ``new``/``delete`` into :proc:`newWithAllocator`/:proc:`deleteWithAllocator`. This
+  new functions allow Chapel code to customize how objects are allocated and
+  deallocated.
+
+  For example, the following code demonstrates how the :type:`mallocWrapper`
+  allocator can be used to allocate a Chapel class.
+
+  .. code-block:: chapel
+
+     class MyClass { var x: int; }
+     var alloc = new mallocWrapper(1024);
+     var x = newWithAllocator(alloc, unmanaged MyClass, 1);
+     writeln(x);
+     deleteWithAllocator(alloc, x);
+
+  Custom allocators can be defined as classes or records that implement the
+  ``allocator`` interface. The interface has two methods: ``allocate`` and
+  ``deallocate``. The ``allocate`` method accepts an integer ``n`` and returns
+  a :type:`~CTypes.c_ptr` to the allocated memory. The ``deallocate``
+  method accepts a :type:`~CTypes.c_ptr` to the allocated memory. Allocators are
+  free to implement their own memory management strategies.
+
+  Limitations:
+    * The ``newWithAllocator`` and ``deleteWithAllocator`` procedures are meant
+      to be stand-ins pending a more elegant syntax to use custom allocators.
+    * This module currently only supports allocating Chapel classes. In the
+      future we hope to support other heap objects like arrays.
+    * Allocating managed (:type:`~OwnedObject.owned`/:type:`~SharedObject.shared`)
+      Chapel classes is not supported, only ``unmanaged`` classes are supported.
+
+*/
+@unstable(category="experimental", reason="The Allocators module is under development and does not have a stable interface yet")
+module Allocators {
+  private use CTypes;
+  private use Reflection;
+  import ChapelLocks;
+  private use ChplConfig;
+
+  private inline proc alignup(ptr: c_ptr(void), param alignment): c_ptr(void)
+    where alignment > 0 && (alignment & (alignment - 1)) == 0 {
+
+    param mask = alignment - 1;
+    const iptr = ptr:c_intptr;
+    const alignedPtr = (iptr + mask) & ~mask;
+    return alignedPtr:c_ptr(void);
+  }
+
+  // TODO: interfaces are not supported by chpldoc yet: https://github.com/chapel-lang/chapel/issues/17383
+  /*
+    All allocators must implement this interface. The interface has two
+    methods: ``allocate`` and ``deallocate``. The ``allocate`` method accepts
+    an integer size and returns a pointer to the allocated memory. The
+    ``deallocate`` method accepts a pointer to the allocated memory.
+
+    Allocators may either be classes or records.
+  */
+  interface allocator {
```

Comment:
IMO these are good candidates for base class with virtual methods and then virtual dispatch. You can list that as possible future work in the PR message.

Reply:
I would prefer them to remain as interfaces and avoid the virtual dispatch, but I agree thats a good candidate for future discussion